File: dotnet/src/Sodium/CSharp/Sodium.Frp.Tests.Internal/TransactionTests.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using NUnit.Framework;
using Sodium.Functional;

namespace Sodium.Frp.Tests.Internal
{
    [TestFixture]
    public class TransactionTests
    {
        [Test]
        public async Task PostSeeOutside()
        {
            OperationCanceledException actual = null;
            AutoResetEvent re = new AutoResetEvent(false);
            using (CancellationTokenSource cts = new CancellationTokenSource())
            {
                Task task = Task.Run(
                    () =>
                    {
                        Transaction.Post(
                            () =>
                            {
                                re.Set();

                                Thread.Sleep(5000);

                                cts.Token.ThrowIfCancellationRequested();
                            });
                    });

                re.WaitOne();

                cts.Cancel();

                try
                {
                    await task;
                }
                catch (OperationCanceledException e)
                {
                    actual = e;
                }
            }

            Assert.IsNotNull(actual);
        }

        [Test]
        public async Task PostSeeInside()
        {
            OperationCanceledException actual = null;
            AutoResetEvent re = new AutoResetEvent(false);
            using (CancellationTokenSource cts = new CancellationTokenSource())
            {
                Task task = Task.Run(
                    () =>
                    {
                        Transaction.Post(
                            () =>
                            {
                                re.Set();

                                Thread.Sleep(5000);

                                cts.Token.ThrowIfCancellationRequested();
                            });
                    });

                re.WaitOne();

                StreamSink<Unit> sink2 = Stream.CreateSink<Unit>();
                sink2.Listen(_ => cts.Cancel());
                sink2.Send(Unit.Value);

                try
                {
                    await task;
                }
                catch (OperationCanceledException e)
                {
                    actual = e;
                }
            }

            Assert.IsNotNull(actual);
        }
    }
}

File: dotnet/src/Sodium/FSharp/Sodium.FSharp.Frp.Tests.Internal/TransactionTests.fs
module Sodium.Frp.Tests.Internal.Transaction

open System
open NUnit.Framework
open Sodium.Frp
open System.Threading

[<TestFixture>]
type ``Transaction Tests``() =

    [<Test>]
    member __.``Post See Outside``() =
        async {
            use re = new AutoResetEvent false
            let! actual =
                async {
                    use cts = new CancellationTokenSource ()
                    let! a =
                        async {
                            Transaction.post (fun () ->
                                re.Set () |> ignore
                                Thread.Sleep 5000
                                cts.Token.ThrowIfCancellationRequested ())
                        } |> Async.StartChild
                    re.WaitOne () |> ignore
                    cts.Cancel ()
                    try
                        do! a
                        return None
                    with
                        | :? OperationCanceledException as e -> return Some e
                }
            Assert.IsTrue (Option.isSome actual)
        } |> Async.StartAsVoidTask

    [<Test>]
    member __.``Post See Inside``() =
        async {
            use re = new AutoResetEvent false
            let! actual =
                async {
                    use cts = new CancellationTokenSource ()
                    let! a =
                        async {
                            Transaction.post (fun () ->
                                re.Set () |> ignore
                                Thread.Sleep 5000
                                cts.Token.ThrowIfCancellationRequested ())
                        } |> Async.StartChild
                    re.WaitOne () |> ignore
                    let sink2 = sinkS ()
                    let _ = (
                        use _l = sink2 |> listenS (fun _ -> cts.Cancel ())
                        sink2 |> sendS ()
                    )
                    try
                        do! a
                        return None
                    with
                        | :? OperationCanceledException as e -> return Some e
                }
            Assert.IsTrue (Option.isSome actual)
        } |> Async.StartAsVoidTask